### PR TITLE
Avoids Shuffling Data Around for Nothing, closes #3306

### LINKED ISSUE #3306: loadNodesFromFile should write into sets directly
We deserialize into vectors but then immediately transform them to sets here:

- https://github.com/Project-OSRM/osrm-backend/blob/538bbd47d1ee169764b66df3ef67ecfe131cdfec/include/util/graph_loader.hpp#L65-L68
- https://github.com/Project-OSRM/osrm-backend/blob/78583d2c8c639466bdde01c848398b2a48b72f44/src/extractor/extractor.cpp#L429-L431

### PR DESCRIPTION
**Pull request by daniel-j-h**
For #3306.

## Tasklist
 - [x] review
 - [x] adjust for comments

### CODE REVIEW

File: src/extractor/extractor.cpp [hunk 1]
```diff
@@ -409,33 +410,31 @@ std::shared_ptr<RestrictionMap> Extractor::LoadRestrictionMap()
...
-                              std::unordered_set<NodeID> &traffic_lights,
+Extractor::LoadNodeBasedGraph(std::unordered_set<NodeID> &barriers,
+                              std::unordered_set<NodeID> &traffic_signals,
                               std::vector<QueryNode> &internal_to_external_node_map)
 {
-    std::vector<NodeBasedEdge> edge_list;
-
-    boost::filesystem::ifstream input_stream(config.output_file_name,
-                                             std::ios::in | std::ios::binary);
+    boost::filesystem::ifstream stream(config.output_file_name, std::ios::binary);
```
Comment (daniel-j-h) [inline src/extractor/extractor.cpp]: ifstream always in

File: src/extractor/extractor.cpp [hunk 2]
```diff
@@ -409,33 +410,31 @@ std::shared_ptr<RestrictionMap> Extractor::LoadRestrictionMap()
...
+    auto barriers_iter = inserter(barriers, end(barriers));
+    auto traffic_signals_iter = inserter(traffic_signals, end(traffic_signals));
 
-    // insert into unordered sets for fast lookup
-    barrier_nodes.insert(barrier_list.begin(), barrier_list.end());
-    traffic_lights.insert(traffic_light_list.begin(), traffic_light_list.end());
+    NodeID number_of_node_based_nodes = util::loadNodesFromFile(stream, //
+                                                                barriers_iter,
+                                                                traffic_signals_iter,
+                                                                internal_to_external_node_map);
```
Comment (daniel-j-h) [inline src/extractor/extractor.cpp]: this is what I wanted to target: no need to create two vectors, transform to sets, clear and shrink vectors again

File: src/tools/components.cpp [hunk 3]
```diff
@@ -53,24 +54,20 @@ std::size_t loadGraph(const char *path,
...
     std::vector<extractor::NodeBasedEdge> edge_list;
-    std::vector<NodeID> traffic_light_node_list;
-    std::vector<NodeID> barrier_node_list;
 
-    auto number_of_nodes = util::loadNodesFromFile(
-        input_stream, barrier_node_list, traffic_light_node_list, coordinate_list);
+    auto nop = boost::make_function_output_iterator([](auto) {});
 
-    util::loadEdgesFromFile(input_stream, edge_list);
+    auto number_of_nodes = util::loadNodesFromFile(stream, nop, nop, coordinate_list);
```
Comment (daniel-j-h) [inline src/tools/components.cpp]: this is a cool trick and a good argument for how powerful iterators are

File: src/extractor/extractor.cpp [hunk 4]
```diff
@@ -409,33 +410,31 @@ std::shared_ptr<RestrictionMap> Extractor::LoadRestrictionMap()
...
 
-    util::SimpleLogger().Write() << " - " << barrier_list.size() << " bollard nodes, "
-                                 << traffic_light_list.size() << " traffic lights";
+    auto barriers_iter = inserter(barriers, end(barriers));
+    auto traffic_signals_iter = inserter(traffic_signals, end(traffic_signals));
 
-    // insert into unordered sets for fast lookup
-    barrier_nodes.insert(barrier_list.begin(), barrier_list.end());
-    traffic_lights.insert(traffic_light_list.begin(), traffic_light_list.end());
+    NodeID number_of_node_based_nodes = util::loadNodesFromFile(stream, //
```
Comment (MoKob) [inline src/extractor/extractor.cpp]: why forcing the `//` line break here?

Comment (daniel-j-h) [inline src/extractor/extractor.cpp]: aesthetics

Comment (daniel-j-h) [inline src/extractor/extractor.cpp]: removed

**Pull request merged**